Add ImplicitFieldSchemaUnsafe (#24113)
microsoft/FluidFramework: .changeset/empty-wasps-smile.md

---
"fluid-framework": minor
"@fluidframework/tree": minor
---
---
"section": tree
---

Improved Type Checking for Recursive Object Schema Fields

Most ways to provide incorrectly typed data for fields of [recursive object schema](https://fluidframework.com/docs/api/fluid-framework/schemafactory-class#objectrecursive-method) now produce simpler type errors without relying on [ValidateRecursiveSchema](https://fluidframework.com/docs/api/fluid-framework/validaterecursiveschema-typealias).

> As a sideeffect of this work some schema which violated the documented allowed patterns specified by [SchemaFactory](https://fluidframework.com/docs/api/fluid-framework/schemafactory-class#schemafactory-remarks) but used to work (as long as they were not package exported) no longer compile.

Reviewer: ```suggestion
As a side effect of this work some schema which violated the documented allowed patterns specified by [SchemaFactory](https://fluidframework.com/docs/api/fluid-framework/schemafactory-class#schemafactory-remarks) but used to work (as long as they were not package exported) no longer compile.
```

> As a side effect of this work some schema which violated the documented allowed patterns specified by [SchemaFactory](https://fluidframework.com/docs/api/fluid-framework/schemafactory-class#schemafactory-remarks) but used to work (as long as they were not package exported) no longer compile.

Reviewer 2: Nit
```suggestion
As a side effect of this work, some schema which violated the documented allowed patterns specified by [SchemaFactory](https://fluidframework.com/docs/api/fluid-framework/schemafactory-class#schemafactory-remarks) but used to work (as long as they were not package exported) no longer compile.
```

The specific case known to break is when:

1. An Object node schema is co-recursive with an Array node schema.

> 2. The Array does not declared a named subclass.

Reviewer: ```suggestion
2. The Array does not declare a named subclass.
```

> 3. The schema reference from the Object to the Array is not using the [lazy syntax](https://fluidframework.com/docs/api/fluid-framework/lazyitem-typealias).

Author: It is using the lazy syntax when the array references the object. It is not doing it when the object references the array.

While the lazy reference in that case is not required (and thus not added in the good example), only the case that doesn't include it happens to break in this change.

For example:

```typescript
class Foo extends sf.objectRecursive("Foo", {
	fooList: sf.arrayRecursive("FooList", [() => Foo]), // Bad
}) {}
{
	type _check = ValidateRecursiveSchema<typeof Foo>;
}
```

Such a schema is disallowed according to the documentation [recursive schema must explicitly declare a named class]((https://fluidframework.com/docs/api/fluid-framework/schemafactory-class#schemafactory-remarks)).

> This restriction is necessary avoid [Generated `.d.ts` files replacing recursive references with `any`](https://github.com/microsoft/TypeScript/issues/55832).

Reviewer: ```suggestion
This restriction is necessary to avoid [generated `.d.ts` files replacing recursive references with `any`](https://github.com/microsoft/TypeScript/issues/55832).
```

Fixing this code is now also necessary to avoid a compile error.

```typescript
// Fixed
class FooList extends sf.arrayRecursive("FooList", [() => Foo]) {}
{
	type _check = ValidateRecursiveSchema<typeof FooList>;
}
class Foo extends sf.objectRecursive("Foo", {
	fooList: FooList,
}) {}
{
	type _check = ValidateRecursiveSchema<typeof Foo>;
}
```

This change will also result in much nicer IntelliSense and type errors while fixing the typing if the schema is exported.

> There are still several cases which still compile but violate this policy regarding recursive schema and can cause issues when exporting schema:

Reviewer 2: Nit ("still" twice)
```suggestion
There are still several cases which compile but violate this policy regarding recursive schema and can cause issues when exporting schema:
```
